clients/src/main/java/org/apache/kafka/clients/consumer/PartitionAssignor.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import org.apache.kafka.clients.consumer.internals.ConsumerProtocol.ConsumerSubscriptionData;
+import org.apache.kafka.clients.consumer.internals.PartitionAssignor.Assignment;
+import org.apache.kafka.clients.consumer.internals.PartitionAssignor.Subscription;
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.SchemaException;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.apache.kafka.clients.consumer.internals.ConsumerProtocol.CONSUMER_PROTOCOL_V0;
+import static org.apache.kafka.clients.consumer.internals.ConsumerProtocol.CONSUMER_PROTOCOL_V1;
+
+/**
+ * This interface is used to define custom partition assignment for use in
+ * {@link org.apache.kafka.clients.consumer.KafkaConsumer}. Members of the consumer group subscribe
+ * to the topics they are interested in and forward their subscriptions to a Kafka broker serving
+ * as the group coordinator. The coordinator selects one member to perform the group assignment and
+ * propagates the subscriptions of all members to it. Then {@link #assign(Cluster, Map)} is called
+ * to perform the assignment and the results are forwarded back to each respective members
+ *
+ * In some cases, it is useful to forward additional metadata to the assignor in order to make
+ * assignment decisions. For this, you can override {@link #subscription(Set)} and provide custom
+ * userData in the returned Subscription. For example, to have a rack-aware assignor, an implementation
+ * can use this user data to forward the rackId belonging to each member.
+ */
+public interface PartitionAssignor {
+
+    /**
+     * Return serialized data that will included in the serializable subscription object and can
+     * be leveraged in {@link #assign(Cluster, Map)} ((e.g. local host/rack information)
+     *
+     * @return Non-null subscription with optional user data
+     */
+    default ByteBuffer userData() {
```

Comment:
I think we need to work into this name the fact that this is the user data which is provided in the join call. I was considering a name like `joinMetadata` or something. Also, we should consider wrapping the result in another type, just in case we need to add anything in the future.

Reply:
Ack on the naming -- but I feel it is ok to leave as ByteBuffer, since this is data which the PartitionAssignor implementation is wholly responsible for. We shouldn't care what it is except to add the serialized bytes to the Subscription, and anything that is added should also be serialized by the PartitionAssignor implementation before being passed up. Does that make sense? @hachikuji

Reply:
Hmm.. I guess the assumption is that there will never be any state that the assignor needs the consumer to be aware of. If such use case did emerge, we could always add another API I suppose (e.g. `subscriptionFoo`). I think I'd still slightly favor wrapping the object (e.g. `SubscriptionMetadata`, but I don't feel too strongly about it.